### ISSUE TITLE
ofi+psm2 protocol low RPC throughput

### ISSUE THREAD

**Issue by carns** [mochi-hpc/mochi-margo#67]
In GitLab by @marcvef on Nov 24, 2020, 06:29

# Issue

I am currently investigating significantly reduced RPC throughput on the latest master branch when the `ofi+psm2` protocol is used between two nodes. Other protocols, e.g., `ofi+sockets` or`ofi+tcp` are unaffected. Interestingly, this issue also appears when `auto_sm` is used on one node with `ofi+psm2`, but does _not_ appear when `auto_sm` is used with `ofi+sockets`, for example. Trying various Margo commits, this issue started to appear in one of the three commits where the initialization functions were refactored 0f115844fd2d8be25f75003a1f265516cb9b1ae9, 143099c4e716b322a6446a4b2ae7f17ad12787eb, 313a5c923e9de018f6dab61cde690fe7656a0045.

# Test

I just quickly ran a few tests to get a grasp on the performance difference with a small MPI program to evaluate raw RPC throughput and bulk_buffer transfer throughput that somewhat resembles GekkoFS' RPC layer between two nodes.

## Setup
We are using two nodes with an 100Gbit Omni-Path Network with either native `ofi+psm2` or TCP over Omni-Path with `ofi+sockets`. Mercury version: v2.0.0, Argobots version v1.0.1, Margo versions before and after aforementioned commits, called `BEFORE` and `AFTER` below. The Margo server uses 8 `rpc_thread_count`

## Raw RPC throughput
Sending 1.6 million RPCs over 16 processes.

| protocol | RPCs/sec (BEFORE) | RPCs/sec (AFTER) |
| ---- | ------ | ------: |
| `ofi+psm2` (remote) | ~90K | **~70** (lower RPC volume) |
| `ofi+sockets` (remote) | ~55K | ~55K |
| `auto_sm (ofi+psm2)` (local) | 100K | **~100** (lower RPC volume) |
| `auto_sm (ofi+sockets)` (local) | 100K | 100K |

## RPC bulk_buffer transfer throughput
Moving ~100 GiB over 16 processes with a bulk buffer of 64 MiB for each RPC in a `BULK_PULL` configuration.

| protocol | MiB/sec (BEFORE) | MiB/sec (AFTER) |
| ---- | ------ | ------: |
| `ofi+psm2` (remote) | ~4300 | **~800** |
| `ofi+sockets` (remote) | ~2200 | ~2200 |
| `auto_sm (ofi+psm2)` (local) | ~11000 | **~2000** |
| `auto_sm (ofi+sockets)` (local) | ~11000 | ~11000 |

My guess is that the reduced data throughput is a side effect due to the reduced RPC throughput in general.

**Comment by carns** [comment]
In GitLab by @carns on Nov 24, 2020, 10:29

Thanks for the detailed report @marcvef !  I'm going to tag in @mdorier and @soumagne on this for input.

Obviously something is different in the before/after that we need to figure out, but one oddity is why ofi+psm2 specifically is affected.  Generally speaking, Margo doesn't particularly care what transport Mercury is using.

The only thing I can think of that's unique about PSM2 is that is probably far more CPU intensive than other transports following this change in Mercury just before the 2.0 release: https://github.com/mercury-hpc/mercury/commit/4410e9790c4c784f4a9a8473124ca8c91011ecec

Possibly there is a bug in Margo that is highlighted more when HG_Progress() is CPU intensive?

@marcvef would it be possible for you to repeat your "AFTER" experiments using Mercury 2.0.0rc3 (this was the last RC before the PSM2 change)?  This might help narrow down if we can confirm that the CPU usage of the ofi+psm2 path is helping to highlight the problem.

Can you also tell us your thread (execution stream) configuration for this test?

**Comment by carns** [comment]
In GitLab by @soumagne on Nov 24, 2020, 14:53

Hmm thinking more about it, with psm2 I think the issue is that margo is not using SEP when using threads. So most likely when busy spinning and having multiple threads the performance is going to drop quite a lot.

**Comment by carns** [comment]
In GitLab by @marcvef on Nov 25, 2020, 10:38

I've ran the same tests with Mercury 2.0.0rc3. When setting `na_init_info.progress_mode = NA_NO_BLOCK` on my side, the throughput is similarly (low). When I don't set it, the throughput is better: around 2x for data, and 5x-10x for raw RPC throughput. But this is the case for Mercury 2.0 as well. So the busy wait of NO_BLOCK seems to affect it.

I also played around with the number of progress threads (`rpc_thread_count`) on the server side with `use_progress_thread` set to `true`. 

It seems, the more threads I use on the server to process RPC calls, the worse the throughput gets (both for RPC and bulk transfers). *But*, if I were to use `-1` for the `rpc_thread_count`, the plain RPC throughput is significantly higher (4x for plain RPC throughput; data throughput stays the same) than in the `BEFORE` case where `8` threads were used. 

There are no further thread (execution stream) settings set on our side. The client uses the following settings: `use_progress_thread` to `false`; `rpc_thread_count` to `0`.

**Comment by carns** [comment]
In GitLab by @mdorier on Nov 25, 2020, 11:18

@carns The refactored init function should in general not affect performance, but there is one thing that I found it does differently when setting up pools:


### Before the refactoring:

ESs created by margo have an `ABT_SCHED_BASIC_WAIT` type of scheduler. The primary ES's scheduler is replaced with an `ABT_SCHED_BASIC_WAIT` scheduler as well, even if it isn't used. The schedulers are created using a call like this:
```
ABT_sched_create_basic(ABT_SCHED_BASIC_WAIT, 0, NULL, ABT_SCHED_CONFIG_NULL, &self_sched);
```
I don't know what a value of 0 for the number of pools mean, it's not documented, but if it means "take over whatever pool were previously associated with the ES", or "make a default one" that means these two ES (primary and progress) will now take work from a `ABT_POOL_FIFO` type of pool.

The RPC ES, however, are associated with a pool that has been initialized with `ABT_POOL_FIFO_WAIT`, and they also have an `ABT_SCHED_BASIC_WAIT` type of scheduler.

### After the refactoring:

All the pools are now `ABT_POOL_FIFO_WAIT` by default (including the primary ES's pool and the progress pool), and all the schedulers are `ABT_SCHED_BASIC_WAIT` by default.

If psm2 doesn't like blocking and context-switching, is it possible that this change causes performance degradation?

Note that we have an easy way of testing that: use `margo_get_config` to spit out the internal state, put that into a JSON file, change "fifo_wait" into "fifo" in the pool definitions, and reinitialize margo using the content of that JSON file.

**Comment by carns** [comment]
In GitLab by @carns on Nov 30, 2020, 09:57

Thanks @marcvef.  Any chance you have a small reproducer we can play with?

It sounds like we need to try a few more things (e.g., toggling the pool type in margo per @mdorier comment to revert to the old (accidental) behavior, or trying NO_BLOCK with other transports to see if the problem is specific to PSM2 or any transport that goes into NO_BLOCK mode).

**Comment by carns** [comment]
In GitLab by @marcvef on Dec 2, 2020, 08:40

Yes, will do, @carns. I can't promise it for this week, but I'll try to get a small reproducer ready for beginning of next week.

**Comment by carns** [comment]
In GitLab by @marcvef on Dec 15, 2020, 11:35

@carns please find a small reproducer [here](https://gitlab.rlp.net/vef/margotester/). It should also be easily modifiable if you'd like to test some configurations.

Building:
```bash
mkdir build && cd build
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=/home/vef/gekkofs_deps/install ..
make
```
Use the `CMAKE_PREFIX_PATH` to point CMake where Argobots, Mercury, and Margo is installed.

Server start with psm2 which will output the address the server is listening on, similar to the margo examples:
```bash
./margotester/build/src/server/rpc_server -P ofi+psm2
```
You can then start the client, e.g., like this:
```bash
mpiexec -np 16 --mca mtl ^psm2,ofi ./margotester/build/src/client/rpc_client -I 1000 -T ofi+psm2://2680b02:0
```
which sends 1000 RPCs per process without bulk transfer to the server, or:
```bash
mpiexec -np 16 --mca mtl ^psm2,ofi ./margotester/build/src/client/rpc_client -I 10 -D -S 65536 -T ofi+psm2://2680b02:0
```
which sends 10 RPCs per process with a bulk transfer of 64 MB each. The client outputs a small performance evaluation for RPC and data throughput (see Readme). Please let me know if there are any questions or difficulties!

**Comment by carns** [comment]
In GitLab by @carns on Dec 22, 2020, 11:36

Thanks @marcvef for the reproducer!

Tagging in @soumagne and @chuckcranor (tying this into the recent conversation about PSM (rather than PSM2) performance issues).  In Marc's test case there are no threads on the client, but the server is using a dedicated progress thread and RPC handler threads (execution streams).

Even though this is a different transport, and it appears to be a server-side issue rather than client-side issue, I wonder if there is a similar root cause where a spin synchronization construct is causing the progress thread to dominate access to a data structure so it's having a hard time processing operations from other threads until a forced kernel time slice (and thus a low RPC rate).

(Unfortunately I haven't had a chance to run the reproducer yet; at this point I likely won't get to look at this closer until after the holiday break)

**Comment by carns** [comment]
In GitLab by @chuckcranor on Dec 22, 2020, 15:07

It is possible.  PSM is a polling-only interface and I think PSM2 is the same.  The spinning causes problems.

I've found that randomly inserting sched_yield() calls can work around some of the issues with the PSM Mercury plugin I've written (I'm no longer using libfabric).

Attached is a simple patch to mercury_core.c that adds a sched_yield() call after the call to NA_Progress() if it returns a NA_TIMEOUT.  If you've got time, give it a try.
[hg.diff](/uploads/e57d2acfc70c74132ab5dace282dce93/hg.diff)

**Comment by carns** [comment]
In GitLab by @chuckcranor on Jan 21, 2021, 12:27

@marcvef Jerome just committed https://github.com/mercury-hpc/mercury/commit/28e3e569ed018ae99c339e16df8459c29f6fa464 ... it might help.

**Comment by carns** [comment]
In GitLab by @carns on Jan 21, 2021, 19:12

Out of band @marcvef reported that he's still seeing the problem.  I'm trying his test case on the Bebop machine at ANL to see if I can reproduce it there, using current master versions of each of margo, mercury, and argobots, and the latest release of PSM2 (I just opened a PR to bump the PSM2 version in spack).

This is the script that I am trying to run right now. I'll report what I find once it runs:

```
#!/bin/bash
#SBATCH -N 2
#SBATCH -A startup-carns
#SBATCH --ntasks-per-node=8
#SBATCH --time=30:00
#SBATCH -p bdwall


export LMOD_PAGER=none

. /home/carns/working/src/spack/share/spack/setup-env.sh

spack load -r mochi-margo
spack load -r boost

spack find --loaded

# make sure that MPI and libfabric can share PSM2
export PSM2_MULTI_EP=1

echo " ==== running server"
srun -N 1 -n 1 --exclusive /home/carns/working/dbg/marc-vef-margo-2021-01/margotester/build/src/server/rpc_server -P ofi+psm2 -H /home/carns/working/dbg/marc-vef-margo-2021-01/hosts &

sleep 5

echo contents of hosts file:
cat /home/carns/working/dbg/marc-vef-margo-2021-01/hosts

srun -N 1 -n 16 --exclusive /home/carns/working/dbg/marc-vef-margo-2021-01/margotester/build/src/client/rpc_client -I 100000 -H /home/carns/working/dbg/marc-vef-margo-2021-01/hosts
```

**Comment by carns** [comment]
In GitLab by @carns on Jan 21, 2021, 19:13

(the above case is a little simplified; one daemon on one node, and 16 client processes on another, but I think that would be sufficient?)

**Comment by carns** [comment]
In GitLab by @carns on Jan 21, 2021, 21:29

I had to back off to a lower number of iterations for the job to complete in 30 minutes.  Here is an example on Bebop that completed that also shows what package versions I used:

```
cat slurm-1971215.out
-- linux-centos7-broadwell / gcc@10.2.0 -------------------------
argobots@main
boost@1.75.0
bzip2@1.0.6
json-c@0.13.1
libfabric@1.11.1
mercury@master
mochi-margo@master
numactl@2.0.11
opa-psm2@11.2.185
zlib@1.2.7
 ==== running server
[2021-01-21 18:51:22.818] [main] [info] main() Initializing environment
[2021-01-21 18:51:22.818] [main] [info] init_environment() Initializing RPC server: 'ofi+psm2://bdw-0049'
[2021-01-21 18:51:23.055] [main] [info] init_rpc_server() Accepting RPCs on address ofi+psm2://4340502:0
contents of hosts file:
bdw-0049 ofi+psm2://4340502:0
Running '16' ranks with '1000' RPCs each: total '16000'. Data enabled 'false'
Starting run()
Finished run() all
----------------RUNTIME----------------
28400106717ns	28400.1ms
563.378 RPCs per second
```

That's definitely too slow.  When I pick this up again I'll try to confirm better performance with an older version of Margo.

**Comment by carns** [comment]
In GitLab by @marcvef on Jan 22, 2021, 05:11

@chuckcranor, thanks for notifying me, and @carns thanks for running the tests! Yes, the test is sufficient and it looks like I am in the same ballpark w.r.t. the throughput you measured. I've also ran some more test with the latest master that includes the tests. Unfortunately, it does not seem to have any impact on the issue and throughput in general (even with the older Margo versions). 

Below are some more results with the latest Mercury master branch.
Used dependencies:
- Mercury@v2.0.1rc1
- Margo@a1f7a24daf319c6946e00fbec7965cb0f2f05cdf (called `new`)
- Margo@0e8ceee9cbc21b08be2b7feadf227768992f72ad (called `old`)
- libfabric@v1.11.1
- opa-psm2@PSM2_11.2.185

I've also ran the server on one node and the client on the other with 16 client processes. With `new` Margo:
```
vef@x0004 ~ $ mpiexec -np 16 --mca mtl ^psm2,ofi ~/vef_m2/margotester/build/src/client/rpc_client -I 1000 -H ~/vef_m2/rpc_hostfile
Running '16' ranks with '1000' RPCs each: total '16000'. Data enabled 'false'
Starting run()
Finished run() all
----------------RUNTIME----------------
34374526753ns   34374.5ms
465.461 RPCs per second
```
compared to `old` Margo which reported `87438.7 RPCs per second` with `-I 100000`.

Data experiment (100 RPCs per process, 64 MiB bulk size for each RPC) with `new` Margo:
```
vef@x0004 ~ $ mpiexec -np 16 --mca mtl ^psm2,ofi ~/vef_m2/margotester/build/src/client/rpc_client -I 100 -H ~/vef_m2/rpc_hostfile -D -S 65536
Running '16' ranks with '100' RPCs each: total '1600'. Data enabled 'true'
Datavolume: '67108864' bytes ('64' MB) per RPC. '102400' MB in total for all ranks
Starting run()
Finished run() all
----------------RUNTIME----------------
114795600476ns  114796ms
13.9378 RPCs per second
892.02 MB per second
```
compared to `old` Margo which reported `4358.01 MB per second` with the same workload.

**Comment by carns** [comment]
In GitLab by @soumagne on Jan 22, 2021, 19:21

I might have been able to finally track it down. I have not been able to verify that same test case but I discovered that when we switched our timers to using CLOCK_MONOTONIC_COARSE, I had also set it to pthread_condattr_setclock() and it looks like it is not well supported. It causes pthread_cond_timedwait() to return far too early, causing extra spin. For apps that do not pass any timeout to HG_Trigger(), that is obviously not an issue but I believe for margo that is one. In any case, could you please give another try to latest mercury master? Thanks.

**Comment by carns** [comment]
In GitLab by @marcvef on Jan 22, 2021, 20:13

Hi Jerome, thanks for spending the time trying to solve this! Unfortunately, the latest Mercury commit does not have any affect on this issue. :/ I just quickly ran the same experiments as in my last comment and the throughputs with the latest Mercury master are similar to the results from above.

**Comment by carns** [comment]
In GitLab by @carns on Jan 24, 2021, 10:38

Ok, this is weird.  I just rebuilt everything with today's (Jan 24) master versions of argobots, mercury, and margo and compared that to a build with a build using an identical environment except that I reverted margo all the way back to 0.7.2 (the last version prior to any json integration).  I was hoping to get one "slow" and one "fast" configuration to compare a couple of things.

The test case has a single server daemon on one node and 16 client processes on another node running 10,000 iterations (this seems sufficient to show the perf problem and runs long enough that I'm not too worried about noise).

Here are the results:
```
-- linux-centos7-broadwell / gcc@10.2.0 -------------------------
argobots@main
boost@1.75.0
bzip2@1.0.6
json-c@0.13.1
mercury@master
mochi-margo@master
zlib@1.2.7
[2021-01-24 08:54:21.363] [main] [info] main() Initializing environment
[2021-01-24 08:54:21.363] [main] [info] init_environment() Initializing RPC server: 'ofi+psm2://bdw-0434'
[2021-01-24 08:54:21.554] [main] [info] init_rpc_server() Accepting RPCs on address ofi+psm2://4b80502:0
Running '16' ranks with '10000' RPCs each: total '160000'. Data enabled 'false'
Starting run()
Finished run() all
----------------RUNTIME----------------
286023477751ns	286023ms
559.395 RPCs per second
```

```
-- linux-centos7-broadwell / gcc@10.2.0 -------------------------
argobots@main
boost@1.75.0
bzip2@1.0.6
libfabric@1.11.1
mochi-margo@0.7.2
numactl@2.0.11
opa-psm2@11.2.185
zlib@1.2.7
[2021-01-24 09:04:45.552] [main] [info] main() Initializing environment
[2021-01-24 09:04:45.553] [main] [info] init_environment() Initializing RPC server: 'ofi+psm2://bdw-0125'
[2021-01-24 09:04:45.761] [main] [info] init_rpc_server() Accepting RPCs on address ofi+psm2://47c0502:0
Running '16' ranks with '10000' RPCs each: total '160000'. Data enabled 'false'
Starting run()
Finished run() all
----------------RUNTIME----------------
540061842123ns	540062ms
296.262 RPCs per second
```

Neither is good, but the margo 0.72 run is actually even slower (I expected this to be the fast configuration).

I guess I need an intermediate version closer to right before the commit that @marcvef pointed out to see good performance?

**Comment by carns** [comment]
In GitLab by @chuckcranor on Jan 24, 2021, 13:02

@carns you posted earlier that mercury na_ofi.c config of ofi+psm2 changed (e.g. went from FI_PROGRESS_AUTO to FI_PROGRESS_MANUAL and NA_OFI_WAIT_FD changed in the define for NA_OFI_PROV_TYPES).  would reverting that make a difference?  depending on the config, the libfabric provider could be forking a spinning thread for each wait operation (e.g. in psmx2_wait_wait() ) which could be bad (especially if you are bound to a cpu core and have to share it with a spinning pthread).

**Comment by carns** [comment]
In GitLab by @carns on Jan 24, 2021, 13:32

Maybe @chuckcranor , I'll try some combos with different revisions of Mercury too.

In the mean time, I modified my job script to shut down the server correctly (so it will call finalize) and modified the json to enable diagnostics.  Here is the diagnostic output from the server:

```
# Margo diagnostics
#Addr Hash and Address Name: 18446744052842898388,ofi+psm2://47d0502:0
# Sun Jan 24 11:24:00 2021

# Function Name, Average Time Per Call, Cumulative Time, Highwatermark, Lowwatermark, Call Count
trigger_elapsed,0.000019789,223.213957787,0.000000477,0.025913954,11279460
progress_elapsed_zero_timeout,0.000001065,11.667953730,0.000000477,0.091005802,10959393
progress_elapsed_nonzero_timeout,0.103065049,6.905358315,0.062404633,0.108000278,67
bulk_create_elapsed,0.000000000,0.000000000,0.000000000,0.000000000,0
```

(FYI, the low watermark and high watermark are inverted; think of those as min and max values here).  

Anyhow, the important thing is that margo (because of the continuous concurrent load from the clients) is staying almost exclusively in a mode where it is calling HG_Progress() with a zero timeout in this test.  There are 67 calls with a non-zero timeout, which probably corresponds roughly to the 5 second sleep before starting the measurement in my script plus a little cleanup time at the end.

Margo never uses a timeout on HG_Trigger().  So whatever is going on is not caused by how Mercury is handling non-zero timeouts in either the progress or trigger code path.

Nothing unusual in this diagnostic, just ruling some things out.

There may be an issue with how margo is handling the progress loop at the margo level, or there could be some problem preventing Mercury from advancing progress on psm2 as expected when called with a zero timeout.

**Comment by carns** [comment]
In GitLab by @carns on Jan 24, 2021, 13:33

Also I opened up a separate ticket because I couldn't get the env variable to enable that diagnostic output to work https://xgitlab.cels.anl.gov/sds/margo/-/issues/75.  It works fine if enabled via json though.

**Comment by carns** [comment]
In GitLab by @carns on Jan 24, 2021, 14:19

Ok, I've tried four combinations now:

*  margo@master mercury@master
*  margo@master mercury@2.0.0rc3
*  margo@0.7.2 mercury@master
*  margo@0.7.2 mercury@2.0.0.rc3

Some of them vary a bit, but none of them are clearing 600 RPC/s.  Mercury@2.0.0rc3 is new enough that it supports the API hooks used by margo master, but old enough that it does not have the wait strategy change for libfabric/psm2.

There should be plenty of cores available; these are haswell nodes with 36 cores each, and the server is configured to use 10 threads total (main, 1 for progress, 8 for rpc handling).  I'll see if I can find an easy way to sanity check the core affinity, though.  It's possible that I am misunderstanding the core bindings and accidentally constraining the server more than I thought?

**Comment by carns** [comment]
In GitLab by @carns on Jan 24, 2021, 14:46

Or maybe I shouldn't assume the problem is on the server side?

**Comment by carns** [comment]
In GitLab by @chuckcranor on Jan 24, 2021, 14:53

FWIW, the cpu_spinlock issue we had with PSM showed up on the client side.  Server was ok.

**Comment by carns** [comment]
In GitLab by @carns on Jan 24, 2021, 15:03

Well if nothing else the test case can be simplified further; I just ran with one client process (instead of 16) one one of the above builds and got 35 RPC/s.  No concurrency needed to show the problem in this test case :)

**Comment by carns** [comment]
In GitLab by @marcvef on Jan 24, 2021, 15:04

@carns this is indeed very odd. Due to cluster maintenance I can't test it right now, but I would expect 0.72 to have good performance because my used version for the "faster" throughput was newer than 0.72 (see commit hash above). I'll try 0.72 asap and report back. To verify that you do not face another issue entirely, you could try to run the margotester with `-P ofi+sockets`, coupled with `-l ib0` (ib0 is in my case the Omni-Path network device). This results in around 50k RPCs/second in my case (regardless of the used Margo version).

**Comment by carns** [comment]
In GitLab by @carns on Jan 24, 2021, 18:03

Alright, finally on to something, just focusing on parameters in the margo@master, mercury@master case.

If I repeat my normal run (16 clients, 10,000 iterations) and just change the server such that the rpc_thread_count is -1, then the RPC rate jumps all the way up to over 172,000 RPCs/s.  Now we are cooking!

That change makes it so that all of the RPC handler ULTs run on the same execution stream (i.e., pthread) as the margo progress thread.  This isn't a solution, but a clue.  It should be perfectly legal to use separate OS threads for RPC handlers, and that if in fact necessary for some services.

This also explains why we haven't seen a problem in the bebop nightly regression scripts; they all operate in a mode that does not use dedicated threads for RPC handling.

This suggests that the progress ULT will safely yield to RPC ULTs (that presumably want to call HG_Respond()), but only if they are in the same OS thread.  If they are in separate OS threads then they are being starved?

**Comment by carns** [comment]
In GitLab by @carns on Jan 24, 2021, 20:35

Sorry about the noise on this ticket, trying not to forget where we left off.

I just confirmed the above behavior by inserting a sched_yield() call immediately after the ABT_thread_yield() call at https://xgitlab.cels.anl.gov/sds/margo/-/blob/master/src/margo-core.c#L1391 .  With that hack in place I can run @marcvef 's test case with @master version of everything, with 8 dedicated RPC threads, and get full rate (172,000+ RPCs/s).

A sched_yield() should not be necessary right there, though.  It looks like for some reason calling HG_Progress() in a tight loop with zero timeout there is starving out other pthreads from issuing HG operations.

**Comment by carns** [comment]
In GitLab by @chuckcranor on Jan 24, 2021, 21:06

So is this is a variation of the hg.diff I posted on this ticket for Marc to try last month?

https://xgitlab.cels.anl.gov/sds/margo/uploads/e57d2acfc70c74132ab5dace282dce93/hg.diff

The PSM progress function uses the PSM ipeek operation to poll for incoming work.  I believe that can be done totally in userspace (everything is mmap()'d into user memory, so no need to make a system call that might context switch the thread like what you would have to do with a TCP socket).   Maybe this allows Mercury with PSM to spin more than it would with something like BMI?

**Comment by carns** [comment]
In GitLab by @carns on Jan 24, 2021, 21:36

Yeah I think that's a good theory @chuckcranor .  BMI will at least make 1 system call per progress iteration no matter how aggressive the locking is elsewhere in the progress path.

**Comment by carns** [comment]
In GitLab by @marcvef on Jan 25, 2021, 07:03

Just to quickly report back. This works for me as well on the Margo and Mercury master. 
- Using `-1` for the rpc_thread_count, I get over 550K RPC/s and the full bandwidth of ~4400 MiB/s with bulk transfers. 
- When I use `8` for the rpc_thread_count with `sched_yield()`, I get around 320K RPC/s, but bulk transfers only result in 800 MiB/s (not sure if this is to be expected).

But the raw RPC throughput is definitely much higher than the 90K that I've seen with the `old` Margo version.

**Comment by carns** [comment]
In GitLab by @carns on Jan 25, 2021, 07:54

I'm not sure what to expect on the multithreaded bulk transfer rate.  You might want to try using numactl (something like `numactl -N 0 -m 0`, but you could try other sockets too) to see what, if any, of that rate dropoff is from thread memory crossing numa domains.

**Comment by carns** [comment]
In GitLab by @marcvef on Jan 25, 2021, 09:19

Good point! This definitely helps with the `8` rpc_thread_count case, pinning the server to either of the two available sockets results in ~4400 MiB/s. Raw RPC throughput, on the other hand, goes down to 110K RPC/s if pinned (vs over 300K RPC/s when not pinned).

Reducing the number of client processes in my specific example can help as well in the unpinned case. More than 4 client processes will reduce the total achieved bandwidth.

**Comment by carns** [comment]
In GitLab by @chuckcranor on Jan 25, 2021, 16:04

@carns So I confirmed the PSM ipeek system call behavior with strace.   I've got a simple PSM server app that waits indefinitely for network progress using ipeek.

The core of the program is this loop:
<pre>
  /* collect async op results */
  pending = 1;  
  loops = 0;
  while (pending > 0) {
    loops++;
    err = psm_mq_ipeek(mq, &gotreq, NULL);
    if (err == PSM_MQ_NO_COMPLETIONS)
      continue;
    if (err != PSM_OK) {
      printf("psm_mq_ipeek: ERR %s\n", psm_error_get_string(err));
    } else {
      printf("psm_mq_test: testing %p\n", gotreq);
      // DO SOMETHING WITH COMPLETE EVENT
      pending--;
    }
  }
</pre>

So it is mainly going to be looping calling psm_mq_ipeek() which is always going to return PSM_MQ_NO_COMPLETIONS (since I'm not going to run the client program in this test).

Result:
<pre>
h0:src/psm-testprog % ./p-rcv &
[2] 7390
h0:src/psm-testprog % psm_init: OK, version is 1.16
psm_ep_open: OK, epid=100000000370303
psm_mq_init: OK
psm_ep_connect: OK: Success 0x969228
psm_epaddr_setctxt: context set worked!
psm_mq_irecv: OK req 0x7ff310755ef0

h0:src/psm-testprog % strace -p 7390
strace: Process 7390 attached
^Cstrace: Process 7390 detached
h0:src/psm-testprog %
</pre>

So no system calls reported by strace while in that loop.  It is a pure CPU spin.

Furthermore, I used gdb to print the value of "loops" while the strace was running:
<pre>
gdb ./p-rcv
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-115.el7
Copyright (C) 2013 Free Software Foundation, Inc.
Reading symbols from /proj/TableFS/data/chuck/src/psm-testprog/p-rcv...done.
(gdb) attach 7390
Attaching to program: /proj/TableFS/data/chuck/src/psm-testprog/./p-rcv, process 7390
...
^C
Program received signal SIGINT, Interrupt.
...
(gdb) up
#3  0x0000000000400d8b in main (argc=<optimized out>, argv=<optimized out>)
    at p-rcv.c:120
120	    err = psm_mq_ipeek(mq, &gotreq, NULL);
(gdb) print loops
$1 = 1161587419
(gdb) 
</pre>

so you can see that it made it through the loop 1161587419 times without making a system call.

The only thing that is going to stop that thread from running is an involuntary context switch due to running out the CPU time quantum.  I believe that PSM2 will have the same behavior... where as something socket-based like BMI will be making system calls in the loop.

Seems like the PSM problems I was working on last year and the PSM2 issue noted in this ticket are related and revolve around polling network interfaces that bypass the kernel.

**Comment by carns** [comment]
In GitLab by @carns on Jan 27, 2021, 09:30

That's interesting.  So it's possible that Mercury's data structures are technically right, but ofi/psm2 itself will starve a thread that's trying to do a (for example) fi_tsend() if the another thread has a tight enough loop polling the completion queue.

I wonder if there are any PSM2 or libfabric settings that would influence this if so.

**Comment by carns** [comment]
In GitLab by @carns on Jan 28, 2021, 13:25

I just looked around briefly at what libfabric's psm2 provider is doing for locking.  The most relevant part for us is probably the completion queue management (https://github.com/ofiwg/libfabric/blob/master/prov/psm2/src/psmx2_cq.c).  Everything is protected with a "psmx2 lock" (https://github.com/ofiwg/libfabric/blob/master/prov/psm2/src/psmx2.h#L892 and https://github.com/ofiwg/libfabric/blob/master/prov/psm2/src/psmx2.h#L924), which in turn maps to a libfabric "fast lock", which itself in turn maps to a pthread spinlock (https://github.com/ofiwg/libfabric/blob/master/include/ofi_lock.h#L59).

This confirms that the libfabric psm2 provider is indeed effectively using spinlocks for all of it's serialization, which likely isn't helping this margo issue.  We've had other problems in libfabric under heavy thread contention due to overly-aggressive spin locking (https://github.com/ofiwg/libfabric/issues/5244).

It would be painstaking to track down and confirm if that is the culprit, but I noticed in their configure script that they have a fallback in case spinlocks are not available on a platform (https://github.com/ofiwg/libfabric/blob/master/configure.ac#L172).  It will globally turn all of the "fast locks" into normal mutexes rather than spinlocks.

I'm curious what would happen if it had an explicit configure option that could be used to turn off the spinlocks (even if they are present), that we could in turn set in spack.  For our use cases I would love to be able to rule out potential libfabric spinlock issues at least as an experiment (and also see exactly what performance improvement they offer in the first place).

I don't have time to try that right now but I wanted to record the thought here.

**Comment by carns** [comment]
In GitLab by @carns on Jan 28, 2021, 13:27

As a side note, I still don't like the term `fast_lock`.  It implies to developers that the only thing special about the lock is it's speed, but spinlocks require more careful handling than normal mutex locks, especially with oversubscribed threads.

**Comment by carns** [comment]
In GitLab by @chuckcranor on Jan 28, 2021, 14:14

yikes, spin overload!  How about just changing the psmx2 lock impl. to always use normal mutexes?  that might be easier to test than messing with configure?

**Comment by carns** [comment]
In GitLab by @carns on Jan 28, 2021, 14:22

I just want a configure option as a debugging aid for all future suspicions of libfabric spinlock problems :)

Replacing them in just psm2 would theoretically not be hard; it looks like function pointers are being used for almost all of them.

**Comment by carns** [comment]
In GitLab by @carns on Feb 2, 2021, 11:35

Well, that experiment didn't go according to plan.  I have a spack package that can build a stack with all libfabric spinlocks replaced with mutex locks (see https://xgitlab.cels.anl.gov/sds/sds-repo/-/blob/master/packages/libfabric/package.py and build your stack with a ^libfabric@master+disable-spinlocks dependency), but when I try to run it, the clients all fail with this message:

```
libfabric:1533:psm2:av:psmx2_epid_to_epaddr():230<warn> psm2_ep_connect retured error Operation timed out, remote epid=1b50502.If it is a timeout error, try setting FI_PSM2_CONN_TIMEOUT to a larger value (current: 5 seconds).
```

I don't know if this means that my hack to replace the spinlocks is faulty, that I was wrong to assume they could be safely replaced, or if it has exposed a locking problem in the code.

**Comment by carns** [comment]
In GitLab by @carns on Feb 9, 2021, 13:30

One more follow up on that experiment.  My previous attempt at running libfabric without spinlocks had a build problem (which @roblatham00 recently caught and fixed).  I got a correct run in today (everything completed fine) but it did not affect performance.

In summary the spinlocks in the libfabric psm2 aren't the culprit this time.  Even with regular mutexes a thread spinning on progress can starve out other threads.

I'm out of ideas other than adding a yield in the path somewhere when the psm2 transport is being used.

**Comment by carns** [comment]
In GitLab by @carns on Feb 26, 2021, 16:35

This issue was addressed at the Mercury level in https://github.com/mercury-hpc/mercury/commit/3c63a13d3c1d9b597552026ca0a251afc0be8e88

**Comment by carns** [comment]
In GitLab by @carns on Feb 26, 2021, 16:35

closed
